NIFI-9861: Removed BlockListClassLoader in favor of AllowListClassLoader

markap14 opened the pull request:
<!--
  Licensed to the Apache Software Foundation (ASF) under one or more
  contributor license agreements.  See the NOTICE file distributed with
  this work for additional information regarding copyright ownership.
  The ASF licenses this file to You under the Apache License, Version 2.0
  (the "License"); you may not use this file except in compliance with
  the License.  You may obtain a copy of the License at
      http://www.apache.org/licenses/LICENSE-2.0
  Unless required by applicable law or agreed to in writing, software
  distributed under the License is distributed on an "AS IS" BASIS,
  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
  See the License for the specific language governing permissions and
  limitations under the License.
-->
Thank you for submitting a contribution to Apache NiFi.

Please provide a short description of the PR here:

#### Description of PR

_Enables X functionality; fixes bug NIFI-YYYY._

In order to streamline the review of the contribution we ask you
to ensure the following steps have been taken:

### For all changes:
- [ ] Is there a JIRA ticket associated with this PR? Is it referenced 
     in the commit message?

- [ ] Does your PR title start with **NIFI-XXXX** where XXXX is the JIRA number you are trying to resolve? Pay particular attention to the hyphen "-" character.

- [ ] Has your PR been rebased against the latest commit within the target branch (typically `main`)?

- [ ] Is your initial contribution a single, squashed commit? _Additional commits in response to PR reviewer feedback should be made on this branch and pushed to allow change tracking. Do not `squash` or use `--force` when pushing to allow for clean monitoring of changes._

### For code changes:
- [ ] Have you ensured that the full suite of tests is executed via `mvn -Pcontrib-check clean install` at the root `nifi` folder?
- [ ] Have you written or updated unit tests to verify your changes?
- [ ] Have you verified that the full build is successful on JDK 8?
- [ ] Have you verified that the full build is successful on JDK 11?
- [ ] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](http://www.apache.org/legal/resolved.html#category-a)?
- [ ] If applicable, have you updated the `LICENSE` file, including the main `LICENSE` file under `nifi-assembly`?
- [ ] If applicable, have you updated the `NOTICE` file, including the main `NOTICE` file found under `nifi-assembly`?
- [ ] If adding new Properties, have you added `.displayName` in addition to .name (programmatic access) for each of the new properties?

### For documentation related changes:
- [ ] Have you ensured that format looks appropriate for the output in which it is rendered?

### Note:
Please ensure that once the PR is submitted, you check GitHub Actions CI for build issues and submit an update to your PR as soon as possible.

turcsanyip commented on nifi-stateless/nifi-stateless-bootstrap/src/main/java/org/apache/nifi/stateless/bootstrap/StatelessBootstrap.java:
```diff
@@ -181,58 +166,75 @@ private static BlockListClassLoader createExtensionRootClassLoader(final File na
-            findClassLoaderUrls(classLoader.getParent(), urls);
+    private static void findLoadableClasses(final File file, final Set<String> classNames) throws IOException {
+        final String filename = file.getName();
+        if (filename.endsWith(".jar")) {
+            findClassNamesInJar(file, classNames);
+        } else if (filename.endsWith(".jmod")) {
+            findClassesInJmod(file, classNames);
```
`.class` files are collected in `findLoadableFiles()` (line 227) but they are not handled here. It did not cause any issues in my flows but there can be other use cases when those classes are needed.

turcsanyip reviewed (changes requested):
@markap14 Thanks for fixing this issue!
I tested it with a simple flow in different environments (run test from IntelliJ, Stateless NiFi, Kafka Connect, ExecuteStateless, all with java 8 and 11). It worked for me.

However, I found a possible issue regarding handling of `.class` files. Pls. see my comment inline.

markap14 commented:
Thanks @turcsanyip . That's a good catch. We can't really add the classes found in .class files because we don't know what the fully qualified class name is. We do know if it's found in a .jar or a .jmod file. Good news is that it doesn't really matter, as it's only used to detect classes that are part of Java itself, and Java packages all of their classes uses the appropriate packing mechanisms (.jmod / .jar) - at least in every case I've found. Hard to imagine that they wouldn't.

So I removed the .class from the list of things we add to the 'loadable files'. I think this is fine and addresses the mismatch between the two. I also rebased as there was a conflict with main.

turcsanyip reviewed (approved):
@markap14 That makes sense. +1 LGTM

Merging to main.
